### PR TITLE
Choose Bind address by default for the kubelet Streaming Server

### LINKED ISSUE #80078: sometime kubelet err when disable ipv6
**What happened**:
kubelet cannot run，there is the err
```
docker_service.go:401] Streaming server stopped unexpectedly: listen tcp [::1]:0: bind: cannot assign requested address
```

**Anything else we need to know?**:
if I enable ipv6, it can run success
**Environment**:
- Kubernetes version (use `kubectl version`): 1.13.5
- Cloud provider or hardware configuration: phy
- OS (e.g: `cat /etc/os-release`): centos7.6
- Kernel (e.g. `uname -a`): 5.1
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### LINKED ISSUE #80278: kubelet start fail for Streaming server stopped unexpected
**Environment:**

Kubernetes version (use kubectl version):
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.4", GitCommit:"a87e9a978f65a8303aa9467537aa59c18122cbf9", GitTreeState:"clean", BuildDate:"2019-07-08T08:51:16Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.4", GitCommit:"a87e9a978f65a8303aa9467537aa59c18122cbf9", GitTreeState:"clean", BuildDate:"2019-07-08T08:43:10Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}

OS (e.g. from /etc/os-release):
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

Kernel (e.g. uname -a):
Linux bd01 4.4.185-1.el7.elrepo.x86_64 #1 SMP Wed Jul 10 08:46:17 EDT 2019 x86_64 x86_64 x86_64 GNU/Linux

Others: other service etcd,flannel,kube-api-server,kube-controller-mannager,kube-scheduler all run normally.
my problem is i have 3 nodes,all these three node  are used as master,worker,etcd role;
tow of these can run kubelet properlly,one can not;
the error journalctl log is:
```
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.344056   21912 container_manager_linux.go:286] Creating device plugin mana
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.344063   21912 manager.go:109] Creating Device Plugin manager at /var/lib/
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.344082   21912 state_mem.go:36] [cpumanager] initializing new in-memory st
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.344246   21912 state_mem.go:84] [cpumanager] updated default cpuset: ""
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.344254   21912 state_mem.go:92] [cpumanager] updated cpuset assignments: "
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.344262   21912 state_checkpoint.go:100] [cpumanager] state checkpoint: res
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.344266   21912 state_checkpoint.go:101] [cpumanager] state checkpoint: def
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.344301   21912 server.go:997] Using root directory: /data/k8s/k8s/kubelet
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.344314   21912 kubelet.go:304] Watching apiserver
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.346407   21912 client.go:75] Connecting to docker on unix:///var/run/docke
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.346428   21912 client.go:104] Start docker client with request timeout=10m
7月 18 08:45:27 bd02 kubelet[21912]: W0718 08:45:27.347736   21912 docker_service.go:561] Hairpin mode set to "promiscuous-bri
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.347751   21912 docker_service.go:238] Hairpin mode set to "hairpin-veth"
7月 18 08:45:27 bd02 kubelet[21912]: W0718 08:45:27.347842   21912 cni.go:213] Unable to update cni config: No networks found
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.349589   21912 docker_service.go:253] Docker cri networking managed by kub
7月 18 08:45:27 bd02 kubelet[21912]: E0718 08:45:27.354624   21912 reflector.go:126] k8s.io/kubernetes/pkg/kubelet/kubelet.go:
7月 18 08:45:27 bd02 kubelet[21912]: E0718 08:45:27.354625   21912 reflector.go:126] k8s.io/kubernetes/pkg/kubelet/config/apis
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.354918   21912 docker_service.go:258] Docker Info: &{ID:CBLH:B63G:MQE2:L3A
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.354968   21912 docker_service.go:271] Setting cgroupDriver to cgroupfs
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.355019   21912 kubelet.go:632] Starting the GRPC server for the docker CRI
7月 18 08:45:27 bd02 systemd[1]: kubelet.service: main process exited, code=exited, status=255/n/a
7月 18 08:45:27 bd02 kubelet[21912]: I0718 08:45:27.355037   21912 docker_server.go:59] Start dockershim grpc server
7月 18 08:45:27 bd02 kubelet[21912]: F0718 08:45:27.355196   21912 docker_service.go:412] Streaming server stopped unexpectedl
7月 18 08:45:27 bd02 systemd[1]: Unit kubelet.service entered failed state.
7月 18 08:45:27 bd02 systemd[1]: kubelet.service failed.
7月 18 08:45:32 bd02 systemd[1]: kubelet.service holdoff time over, scheduling restart.
7月 18 08:45:32 bd02 systemd[1]: Started Kubernetes Kubelet.
```
another node which can run kubelet correct,it jouralctl log is :
```
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.728635    8565 container_manager_linux.go:286] Creating device plugin manag
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.728643    8565 manager.go:109] Creating Device Plugin manager at /var/lib/k
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.728662    8565 state_mem.go:36] [cpumanager] initializing new in-memory sta
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.728745    8565 state_mem.go:84] [cpumanager] updated default cpuset: ""
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.728753    8565 state_mem.go:92] [cpumanager] updated cpuset assignments: "m
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.728760    8565 state_checkpoint.go:100] [cpumanager] state checkpoint: rest
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.728764    8565 state_checkpoint.go:101] [cpumanager] state checkpoint: defa
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.728796    8565 server.go:997] Using root directory: /data/k8s/k8s/kubelet
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.728807    8565 kubelet.go:304] Watching apiserver
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.730032    8565 client.go:75] Connecting to docker on unix:///var/run/docker
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.730042    8565 client.go:104] Start docker client with request timeout=10m0
7月 17 10:37:05 bd03 kubelet[8565]: W0717 10:37:05.730872    8565 docker_service.go:561] Hairpin mode set to "promiscuous-brid
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.730881    8565 docker_service.go:238] Hairpin mode set to "hairpin-veth"
7月 17 10:37:05 bd03 kubelet[8565]: W0717 10:37:05.730945    8565 cni.go:213] Unable to update cni config: No networks found i
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.732613    8565 docker_service.go:253] Docker cri networking managed by kube
7月 17 10:37:05 bd03 kubelet[8565]: E0717 10:37:05.733535    8565 reflector.go:126] k8s.io/kubernetes/pkg/kubelet/kubelet.go:4
7月 17 10:37:05 bd03 kubelet[8565]: E0717 10:37:05.734019    8565 reflector.go:126] k8s.io/kubernetes/pkg/kubelet/kubelet.go:4
7月 17 10:37:05 bd03 kubelet[8565]: E0717 10:37:05.734096    8565 reflector.go:126] k8s.io/kubernetes/pkg/kubelet/config/apise
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.736683    8565 docker_service.go:258] Docker Info: &{ID:WOZU:F6UW:H6C4:4SFS
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.736734    8565 docker_service.go:271] Setting cgroupDriver to cgroupfs
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.736796    8565 kubelet.go:632] Starting the GRPC server for the docker CRI
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.736813    8565 docker_server.go:59] Start dockershim grpc server
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.740656    8565 remote_runtime.go:62] parsed scheme: ""
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.740667    8565 remote_runtime.go:62] scheme "" not registered, fallback to
7月 17 10:37:05 bd03 kubelet[8565]: I0717 10:37:05.740683    8565 remote_image.go:50] parsed scheme: ""
```
compare to correct jouralctl log,the difference is ：
```
7月 17 10:37:05 bd03 kubelet[8565]: E0717 10:37:05.733535    8565 reflector.go:126] k8s.io/kubernetes/pkg/kubelet/kubelet.go:4
7月 17 10:37:05 bd03 kubelet[8565]: E0717 10:37:05.734019    8565 reflector.go:126] k8s.io/kubernetes/pkg/kubelet/kubelet.go:4
7月 17 10:37:05 bd03 kubelet[8565]: E0717 10:37:05.734096    8565 reflector.go:126] k8s.io/kubernetes/pkg/kubelet/config/apise
```
and /var/log/messages error i found is below:
```
Jul 18 08:45:37 bd02 kubelet: I0718 08:45:37.720720   22039 docker_service.go:271] Setting cgroupDriver to cgroupfs
Jul 18 08:45:37 bd02 kubelet: I0718 08:45:37.720778   22039 kubelet.go:632] Starting the GRPC server for the docker CRI shim.
Jul 18 08:45:37 bd02 systemd: kubelet.service failed.
Jul 18 08:45:37 bd02 kubelet: I0718 08:45:37.720795   22039 docker_server.go:59] Start dockershim grpc server
Jul 18 08:45:37 bd02 kubelet: F0718 08:45:37.720999   22039 docker_service.go:412] Streaming server stopped unexpectedly: listen tcp [::1]:0: bind: cannot assign requested address
Jul 18 08:45:37 bd02 etcd: request "header:<ID:15954401848648262393 username:\"kubernetes\" auth_revision:1 > txn:<compare:<target:MOD key:\"/registry/certificatesigningrequests/csr-cjqfh\" mod_revision:397146 > success:<request_put:<key:\"/registry/certificatesigningrequests/csr-cjqfh\" value_size:1854 >> failure:<>>" with result "size:18" took too long (116.497963ms) to execute
Jul 18 08:45:39 bd02 systemd: Stopped Kubernetes Kubelet.
Jul 18 08:45:45 bd02 etcd: request "header:<ID:15954401848648262466 username:\"kubernetes\" auth_revision:1 > txn:<compare:<target:MOD key:\"/registry/services/endpoints/kube-system/kube-controller-manager\" mod_revision:397156 > success:<request_put:<key:\"/registry/services/endpoints/kube-system/kube-controller-manager\" value_size:357 >> failure:<>>" with result "size:18" took too long (208.730124ms) to execute
Jul 18 08:45:45 bd02 etcd: read-only range request "key:\"/registry/services/specs/default/kubernetes\" " with result "range_response_count:1 size:295" took too long (887.437011ms) to execute
```
i have spended tow day to find out what happens,but got no result.
so can someone help me?
thks.
/sig network /sig cluster-lifecycle

### PR DESCRIPTION
**Pull request by aojea**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

It picks an usable binding address by defaut for the Streaming Server in kubelet

**Which issue(s) this PR fixes**:

Fixes #80278
Fixes #80078

**Special notes for your reviewer**:

I'm  not sure this is the best approach because I don't know the goal is to listen in a loopback address by default.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
I was digging more into the code 

https://github.com/kubernetes/kubernetes/blob/174b6d0e2fc99d9964a7d5a7484aa0b7d50b4be1/pkg/kubelet/server/streaming/server.go#L238

and I think that the empty address fits better in this situation,
 
>, if the host in the address parameter is empty .... Listen listens on all available unicast and anycast IP addresses of the local system
> If the port in the address parameter is empty or "0",...,  a port number is automatically chosen.

https://golang.org/pkg/net/#Listen

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2291,7 +2291,7 @@ func getStreamingConfig(kubeCfg *kubeletconfiginternal.KubeletConfiguration, kub
 		SupportedPortForwardProtocols:   streaming.DefaultConfig.SupportedPortForwardProtocols,
 	}
 	if !crOptions.RedirectContainerStreaming {
-		config.Addr = net.JoinHostPort("localhost", "0")
+		config.Addr = ""
```
Comment (mattjmcnaughton) [inline pkg/kubelet/kubelet.go]: From your comment, I'm understanding that setting to `""` means that now we are listening on more IP address than just localhost? Is this a security concern?

Is there an option where we still specify `localhost` but do so in a way that doesn't require ipv6?

Regardless of the path forward, it'd be great to have a descriptive variable/function name to make it clearer the purpose of `config.Addr = ""`.

**Review by mattjmcnaughton — COMMENTED** [review]
/test pull-kubernetes-integration

Thanks for your pr here :)

From what I'm reading, it seems like this diff has a behavioral change - i.e. we will start listening on more ips. Just want to make sure that doesn't pose any issues.

**Comment by mattjmcnaughton** [comment]
Also, I'm curious because I've seen it come up in a couple of diffs - what are k8s expectations around ipv6? Is k8s expected to function completely normally if ipv6 is disabled on the hosts running the different k8s components?

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (aojea) [reply pkg/kubelet/kubelet.go]: > From your comment, I'm understanding that setting to "" means that now we are listening on more IP address than just localhost? Is this a security concern?

that's my concern too, I really don't know what's the purpose of using `localhost`, that`s why assigned the PR to the patch author :)

> Is there an option where we still specify localhost but do so in a way that doesn't require ipv6?

The problem is, AFAIK,  this depends on the OS mechanism to resolve `localhost` and its preference. 

> Regardless of the path forward, it'd be great to have a descriptive variable/function name to make it clearer the purpose of config.Addr = "".

Is not the address where the streaming server is listening?

**Comment by aojea** [comment]
> Also, I'm curious because I've seen it come up in a couple of diffs - what are k8s expectations around ipv6? Is k8s expected to function completely normally if ipv6 is disabled on the hosts running the different k8s components?

I think that k8s should work normally in ipv4 only and ipv6 only environments, that are the modes that it supports currently, but maybe is more an architectural question ...

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (thockin) [reply pkg/kubelet/kubelet.go]: The problem seems to be that "localhost" is resolving to an IPv6 address, but IPv6 is disabled, right?

Isn't that just a bad configuration on the machine?

**Comment by thockin** [comment]
I do not think the right answer is to listen on every interface (what this does).  "localhost" is supposed to work.  It's one of those things that underpins everything.

I added some comments on https://github.com/kubernetes/kubernetes/issues/80278 - let's make sure we understand teh problem before trying to fix it.  I am pretty sure this is NOT the right fix..

**Comment by aojea** [comment]
/hold


> I do not think the right answer is to listen on every interface (what this does). "localhost" is supposed to work. It's one of those things that underpins everything.
> I added some comments on #80278 - let's make sure we understand teh problem before trying to fix it. I am pretty sure this is NOT the right fix..


Let me try to explain myself better in the ticket, doesn´t make any sense discuss the same in both places

**Comment by thockin** [comment]
Closing as per #80278 - scream if you still disagree

**Comment by aojea** [comment]
I was going to close it right now, thanks

**Pull request closed without merging**